File: packages/scriba/src/cli/render.test.ts
import { describe, expect, it } from 'vitest'
import { SCHEMA_VERSION } from '../schema/model.ts'
import { renderReport, renderStatus } from './render.ts'

describe('human renderers', () => {
	it('renders status output with high-contrast text and readable progress', () => {
		const rendered = renderStatus({
			schemaVersion: SCHEMA_VERSION,
			generatedAt: '2026-05-05T13:00:00.000Z',
			providers: [
				{
					providerId: 'codex',
					displayName: 'Codex',
					state: 'ok',
					provenance: [],
					lines: [
						{ type: 'badge', label: 'Plan', text: 'prolite' },
						{
							type: 'progress',
							label: '5h limit',
							used: 18,
							limit: 100,
							format: { kind: 'percent' },
							resetsAt: '2099-05-05T18:00:00.000Z',
						},
						{ type: 'text', label: 'Today', value: '23,135,924' },
					],
				},
			],
		})

		expect(rendered).toContain('Codex')
		expect(rendered).toContain('5h limit')
		expect(rendered).toContain('18%')
		expect(rendered).toContain('resets in')
		expect(rendered).not.toContain('used 18% of 100%')
		expect(rendered).toContain('▰▰▰▰')
		expect(rendered).not.toContain('░')
	})

	it('renders report tables without murky placeholder glyphs', () => {
		const rendered = renderReport('Codex Daily', {
			providerId: 'codex',
			stats: {
				files: 2,
				bytes: 1024,
				lines: 4,
				events: 1,
				invalidLines: 0,
				duplicates: 0,
				missingDirectories: [],
			},
			rows: [
				{
					date: '2026-05-05',
					totalTokens: 1200,
					inputTokens: 1000,
					outputTokens: 200,
					cacheReadTokens: 100,
					costUSD: 0.1234,
				},
			],
		})

		expect(rendered).toContain('Codex Daily')
		expect(rendered).toContain('bucket')
		expect(rendered).toContain('2026-05-05')
		expect(rendered).not.toContain('░')
	})
})

File: internal/render/render_test.go
package render

import (
	"regexp"
	"strings"
	"testing"
	"time"

	"github.com/agensfield/scriba/internal/model"
)

func TestCodexLimitsHidesSparkLines(t *testing.T) {
	used := 12.0
	limit := 100.0
	text := CodexLimits([]model.MetricLine{
		{Type: "badge", Label: "Plan", Text: "prolite"},
		{Type: "progress", Label: "5h limit", Used: &used, Limit: &limit},
		{Type: "progress", Label: "Spark 5h", Used: &used, Limit: &limit},
		{Type: "progress", Label: "Spark weekly", Used: &used, Limit: &limit},
	}, false)
	if strings.Contains(text, "Spark") {
		t.Fatalf("expected Spark lines to be hidden:\n%s", text)
	}
	if !strings.Contains(text, "5h limit") || !strings.Contains(text, "Plan") {
		t.Fatalf("expected non-Spark lines to remain:\n%s", text)
	}
}

func TestCodexLimitsShowsResetGrants(t *testing.T) {
	grants := 1.0
	text := stripANSI(CodexLimits([]model.MetricLine{
		{Type: "amount", Label: "Reset grants", Value: grants, Format: &model.MetricFormat{Kind: "count", Suffix: "available"}},
	}, false))

	if !strings.Contains(text, "Reset grants") || !strings.Contains(text, "1 available") {
		t.Fatalf("expected reset grants to render:\n%s", text)
	}
}

func TestCodexLimitsFormatsGrantExpiry(t *testing.T) {
	text := stripANSI(CodexLimits([]model.MetricLine{
		{Type: "text", Label: "Grant expiry", Value: "2026-07-12T01:20:48.728491Z"},
	}, false))

	want := time.Date(2026, 7, 12, 1, 20, 48, 728491000, time.UTC).Local().Format("2006-01-02 15:04 MST")
	if strings.Contains(text, "728491") || !strings.Contains(text, want) {
		t.Fatalf("expected human grant expiry:\n%s", text)
	}
}

func TestStatusHidesSparkLines(t *testing.T) {
	used := 12.0
	limit := 100.0
	text := Status(model.StatusSnapshot{
		SchemaVersion: model.SchemaVersion,
		GeneratedAt:   "2026-06-01T00:00:00Z",
		Providers: []model.ProviderSnapshot{{
			ProviderID:  "codex",
			DisplayName: "Codex",
			Lines: []model.MetricLine{
				{Type: "progress", Label: "5h limit", Used: &used, Limit: &limit},
				{Type: "progress", Label: "Spark weekly", Used: &used, Limit: &limit},
			},
		}},
	})
	if strings.Contains(text, "Spark") {
		t.Fatalf("expected Spark lines to be hidden:\n%s", text)
	}
	if !strings.Contains(text, "5h limit") {
		t.Fatalf("expected 5h line to remain:\n%s", text)
	}
}

func TestStatusSanitizesOAuthJSONErrors(t *testing.T) {
	text := stripANSI(Status(model.StatusSnapshot{
		SchemaVersion: model.SchemaVersion,
		GeneratedAt:   "2026-06-01T00:00:00Z",
		Providers: []model.ProviderSnapshot{{
			ProviderID:  "claude",
			DisplayName: "Claude",
			Provenance: []model.SourceProvenance{{
				Error: `claude OAuth credentials found but refresh failed: refresh failed: 400 {"error":"invalid_grant","error_description":"Refresh token not found or invalid"}`,
			}},
		}},
	}))

	if strings.Contains(text, `"error"`) || !strings.Contains(text, "OAuth refresh failed: Refresh token not found or invalid") {
		t.Fatalf("expected sanitized OAuth error:\n%s", text)
	}
}

func TestReportShowsHumanRows(t *testing.T) {
	cost := 1.25
	text := stripANSI(Report("Codex Weekly", []model.WeeklyReportRow{
		{
			Week: "2026-06-29",
			ReportTotals: model.ReportTotals{
				TokenUsage: model.TokenUsage{InputTokens: 1200, OutputTokens: 300, TotalTokens: 1500},
				CostUSD:    &cost,
			},
			Models: []model.ModelBreakdown{{Model: "gpt-5-codex", TokenUsage: model.TokenUsage{TotalTokens: 1500}}},
		},
	}))

	for _, want := range []string{"Codex Weekly", "1 weeks", "week of 2026-06-29", "1.5K tokens", "$1.25", "gpt-5-codex"} {
		if !strings.Contains(text, want) {
			t.Fatalf("report missing %q:\n%s", want, text)
		}
	}
	if strings.Contains(text, "rows") {
		t.Fatalf("report leaked row-count wording:\n%s", text)
	}
}

func TestCodexLimitsAlignsProgressBars(t *testing.T) {
	fiveUsed := 8.0
	weeklyUsed := 23.0
	limit := 100.0
	text := stripANSI(CodexLimits([]model.MetricLine{
		{Type: "badge", Label: "Plan", Text: "prolite"},
		{Type: "progress", Label: "5h limit", Used: &fiveUsed, Limit: &limit},
		{Type: "progress", Label: "Weekly limit", Used: &weeklyUsed, Limit: &limit},
	}, false))

	lines := strings.Split(text, "\n")
	barColumns := map[string]int{}
	for _, line := range lines {
		if strings.Contains(line, "limit") {
			barColumns[strings.TrimSpace(line[:strings.Index(line, "▰")])] = strings.Index(line, "▰")
		}
	}
	if barColumns["5h limit"] != barColumns["Weekly limit"] {
		t.Fatalf("progress bars are not aligned:\n%s", text)
	}
}

func stripANSI(text string) string {
	return regexp.MustCompile(`\x1b\[[0-9;]*m`).ReplaceAllString(text, "")
}
